Simplified/Reader2/Internal/LibraryService.swift

```diff
@@ -13,245 +13,81 @@
-
+import R2Streamer
+
+final class LibraryService: Loggable {
+  
+  private let streamer: Streamer
+  private let publicationServer: PublicationServer
+  private var drmLibraryServices = [DRMLibraryService]()
+  
+  private lazy var documentDirectory = try! FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
+  
   init(publicationServer: PublicationServer) {
     self.publicationServer = publicationServer
-
+    
     #if LCP
     drmLibraryServices.append(LCPLibraryService())
     #endif
     
     #if FEATURE_DRM_CONNECTOR
-    drmLibraryServices.append(AdobeDRMLibraryService())
+//    drmLibraryServices.append(AdobeDRMLibraryService())
     #endif
+    
+    streamer = Streamer(
+      contentProtections: drmLibraryServices.compactMap { $0.contentProtection }
+    )
   }
-
-  /// Complementary parsing of the publication.
-  /// Will parse Nav/ncx + mo (files that are possibly encrypted)
-  /// using the DRM object of the publication.container.
-  func loadDRM(for book: NYPLBook, completion: @escaping (CancellableResult<DRM?>) -> Void) {
-
-    guard let filename = book.fileName, let fileUrl = URL(string: filename), let (container, parsingCallback) = items[fileUrl.lastPathComponent] else {
-      completion(.success(nil))
-      return
-    }
-
-    guard let drm = container.drm else {
-      // No DRM, so the parsing callback can be directly called.
-      do {
-        try parsingCallback(nil)
-        completion(.success(nil))
-      } catch {
-        completion(.failure(error))
-      }
-      return
-    }
-
-    guard let drmService = drmLibraryServices.first(where: { $0.brand == drm.brand }) else {
-      completion(.success(nil))
+  
+  
+  // MARK: Opening
+  
+  /// Opens the Readium 2 Publication for the given `book`.
+  ///
+  /// If the `Publication` is intended to be presented in a navigator, set `forPresentation`.
+  func openBook(_ book: NYPLBook, sender: UIViewController, completion: @escaping (CancellableResult<Publication, LibraryError>) -> Void) {
+    guard let bookUrl =  book.url else {
+      completion(.failure(.publicationIsNotValid))
       return
     }
-    
-    #if FEATURE_DRM_CONNECTOR
-    // Load DRM service with publication data
-    // If it's Adobe DRM, set contaiener for decrypting
-    if let adobeDrmService = drmService as? AdobeDRMLibraryService {
-      adobeDrmService.container = container
-    }
-    #endif
-
-    let url = URL(fileURLWithPath: container.rootFile.rootPath)
-    drmService.loadPublication(at: url, drm: drm) { result in
-      switch result {
-      case .success(let drm):
-        do {
-          /// Update container.drm to drm and parse the remaining elements.
-          try parsingCallback(drm)
-          completion(.success(drm))
-        } catch {
-          completion(.failure(error))
+    deferredCatching { .success(bookUrl) }
+      .flatMap { self.openPublication(at: $0, allowUserInteraction: true, sender: sender) }
```

Comment:
`Deferred` type resolves on the main queue, `.flatMap` is performed through nested `Deferred` and also resolves on the main queue

Reply:
sounds good. 

but how do you know that? is it because `openBook` is called on the main thread and `deferredCatching` continues to work on the same thread it's called on?

Reply:
`Deferred` class is designed this way, default queue for `resolve` function completion is main queue.
https://github.com/readium/r2-shared-swift/blob/77896c84a21cd618a6ce09634a3d4b0516bbd074/r2-shared-swift/Toolkit/Deferred.swift#L98
(`deferredCatching` returns a new `Deferred`)